README.md
# wp-vimeo-embed

#### Version: 0.3.6

[![Build Status](https://travis-ci.org/ideasonpurpose/wp-vimeo-embed.svg?branch=master)](https://travis-ci.org/ideasonpurpose/wp-vimeo-embed)
[![Coverage Status](https://coveralls.io/repos/github/ideasonpurpose/wp-vimeo-embed/badge.svg?branch=master)](https://coveralls.io/github/ideasonpurpose/wp-vimeo-embed?branch=master)
[![codecov](https://codecov.io/gh/ideasonpurpose/wp-vimeo-embed/branch/master/graph/badge.svg)](https://codecov.io/gh/ideasonpurpose/wp-vimeo-embed)

A collection of Vimeo embed tools and shortcodes for use in WordPress themes.

## Shortcodes

The following shortcodes will be supported:

- `[vimeo 1234567]` -- straight embed, stretches to 100% width
- `[vimeo 1234567 loop]` -- video tag embed, loops
- `[vimeo 1234567 autoplay]` -- video tag embed, autoplay
- `[vimeo 1234567 loop autoplay]` -- video tag embed, loops and autoplays
- `[vimeo 1234567 autoPLAY LoOp]` -- same as above (order and case don't matter)
- `[vimeo 1234567 lightbox]` -- standard embed pops open an Ekko lightbox wrapper

## Code

There are three methods for injecting Vimeo embed codes in to a page:

```php
$vimeo->wrap($vimeoID); // Wrap Vimeo's embed code in a responsive wrapper
$vimeo->embed($vimeoID, [$arg1]); // Inject an HTML5 video tag
$vimeo->lightbox($vimeoID); // Inject a image link which opens a video lightbox
```

The library should be initialized with a Vimeo API token:

```php
use ideasonpurpose/VimeoEmbed;

$vimeo = new VimeoEmbed('a1234a2bbdcc9d43250b2aefcff944ce');

$vimeo->embed('1234567');
```

Or, using the output directly from ACF Pro's oEmbed field:

    $vimeo->wrap(get_field('video'));

## Usage

This library is not on Packagist so Composer needs to be told where to find it. Add this to the `composer.json` `repositories` key:

```json
  "repositories": [
    {
      "type": "vcs",
      "url": "https://github.com/ideasonpurpose/wp-vimeo-embed"
    }
  ]
```

Then tell Composer to load the package:

```
$ composer require ideasonpurpose/wp-vimeo-embed
```

Initialize the code with a a Vimeo API token:

```php
use ideasonpurpose/VimeoEmbed;

$vimeo = new VimeoEmbed('1234567890abcdef0000000000000000');
```

## Development

Install dependencies with `composer install` and `npm install`.

Tests run with [PHPUnit][]. All tests have been wrapped in package.json script commands. To run test suite use `npm test`. To watch all files and automatically re-run the test suite on changes use `npm run watch`.

[phpunit]: https://phpunit.de/


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.3.6](https://github.com/ideasonpurpose/wp-vimeo-embed/compare/v0.3.4...v0.3.6)

- var interpolation syntax fix [`0c8fe8b`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/0c8fe8b80fc723687277dbf847908391e956362e)

#### [v0.3.4](https://github.com/ideasonpurpose/wp-vimeo-embed/compare/v0.3.3...v0.3.4)

> 18 September 2023

- version cleanup [`c26d665`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/c26d66514830e7319f3385d4d96f6a819ca80e75)

#### [v0.3.3](https://github.com/ideasonpurpose/wp-vimeo-embed/compare/v0.3.2...v0.3.3)

> 18 September 2023

- php 8.2 dynamic props deprecations, bump deps [`bd9f490`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/bd9f490e965c8fa14cbd613d6be2bec5c14d7db6)

#### [v0.3.2](https://github.com/ideasonpurpose/wp-vimeo-embed/compare/v0.1.1...v0.3.2)

> 8 September 2020

- Remove double-slash in vimeo-embed enqueue [`#9`](https://github.com/ideasonpurpose/wp-vimeo-embed/pull/9)
- Bump lodash from 4.17.11 to 4.17.20 [`#10`](https://github.com/ideasonpurpose/wp-vimeo-embed/pull/10)
- Update video source url with https secure source [`#8`](https://github.com/ideasonpurpose/wp-vimeo-embed/pull/8)
- bump dependencies, get tests working again [`a2bdeb5`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/a2bdeb54ab60100f7674a423fb54d050ca5c3c08)
- update readme, remove Yarn [`42fd51d`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/42fd51d07d091f7fb6feed99ee8b6875432b461c)
- switch to Guzzle, add docker, update PHPUnit tests [`e841df4`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/e841df4cf1126d19a071e8a2067db14288635d33)
- bump deps, composer, add changelog [`e9b33eb`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/e9b33eb9835dba470c0de083f9bd4e8c04141d0b)
- dependencies and boilerplate [`6b320a3`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/6b320a3b92c79bbda1eec9a597e589a677548fad)
- update VimeoEmbed, add init.js script [`c4b6ddb`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/c4b6ddb8de1eb1ee8eda40982c8b8d9194b9aa71)
- prettier, still a mess [`992c1ea`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/992c1eaf587cf5d9b0665e28cae94615856da654)

#### v0.1.1

> 20 June 2017

- html5 attribute tests and playsinline. Closes #1 [`#1`](https://github.com/ideasonpurpose/wp-vimeo-embed/issues/1)
- updating VimeoEmbed lib [`f6c3da9`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/f6c3da9b62b012479afd27900be527d25ae91844)
- vimeoID tests, updates [`d9cee40`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/d9cee40c323faf5567bcbe0867adff75b3ea9213)
- updates and test structure [`47632a2`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/47632a2bbdcc9d43250b2aefcff944cee3fc0ee3)
- shortcode updates and code coverage [`ce691a5`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/ce691a56b5d8add4069adc5884588c6fbb8c1f27)
- first commit [`dd424a8`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/dd424a8709222b754a91a125d68a06967826b0dc)
- readme update, test Error to global namespace [`23ceea9`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/23ceea9e7fc81bd85ab12f58559295cf53111fb9)
- docs, cleanup [`409a91e`](https://github.com/ideasonpurpose/wp-vimeo-embed/commit/409a91ec3d56b0bfc486fef0f689d81b7c8aa7ff)


tests/TestCase.php
<?php

use Brain\Monkey;

class TestCase extends PHPUnit_Framework_TestCase
{
    protected function setUp()
    {
        parent::setUp();
        Monkey\setUp();
    }

    protected function tearDown()
    {
        Monkey\tearDown();
        parent::tearDown();
    }
}


tests/VimeoEmbedTest.php
<?php

namespace ideasonpurpose;

use PHPUnit\Framework\TestCase;
use Brain\Monkey\Functions;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Exception\RequestException;

// use Requests;

// It should wrap Vimeo's embed code

// It should Create a video tag embed
// it should ignore case of loop and autoplay attributes
//
// it should embed a lightbox link
// t should include javascript files with lightbox
//
// it should make a Vimeo API request
// it should store the API response in a transient
// it should fail on network errors

class VimeoEmbedTest extends TestCase
{
    protected function setUp(): void
    {
        global $stub, $req;
        $stub = $this->getMockBuilder('ideasonpurpose\VimeoEmbed')
            ->disableOriginalConstructor()
            ->setMethods(['apiGet'])
            ->getMock();

        $stub->method('apiGet')->willReturn((object) ['name' => 'vimeo']);

        Functions\when('set_transient')->justReturn(true);
        Functions\when('get_transient')->justReturn(false);
        Functions\when('delete_transient')->justReturn(true);
        Functions\when('add_shortcode')->justReturn(true);
        Functions\when('add_action')->justReturn(true);
        parent::setUp();
    }

    /**
     * Make sure tests are working
     */
    public function test()
    {
        $this->assertTrue(true);
        $this->assertFalse(false);
    }

    /**
     * It should return strings that might be Vimeo IDs
     *
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGetVimeoIdFromVimeoID()
    {
        global $stub;

        $this->assertEquals('123456', $stub->getVimeoData(123456)->id);
        $this->assertEquals('123456', $stub->getVimeoData('123456')->id);
        $this->assertEquals('notAnId', $stub->getVimeoData('notAnId')->id);
    }

    /**
     * It should extract a Vimeo ID from oEmbed code blobs
     *
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGetVimeoIdFromEmbedCode()
    {
        global $stub;
        $oEmbedBlob =
            '<iframe src="https://player.vimeo.com/video/216711407" width="519" height="390" frameborder="0" title="Navigators 2016 Digital Annual Report" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>';
        $this->assertEquals('216711407', $stub->getVimeoData($oEmbedBlob)->id);

        $oEmbedBlob =
            '<iframe src="https://player.vimeo.com/video/2822787?color=ffffff&title=0&portrait=0" width="640" height="360" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>';
        $this->assertEquals('2822787', $stub->getVimeoData($oEmbedBlob)->id);
    }

    /**
     * It should throw an exception with bad data and WP_DEBUG set
     *
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGetVimeoIdFailWithException()
    {
        define('WP_DEBUG', true);
        global $stub;
        $this->expectException(\Exception::class);
        $stub->getVimeoData(null);
        $stub->getVimeoData('');
    }

    /**
     * It should return the error wrapped in a comment
     *
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testGetVimeoIdFailWithComment()
    {
        global $stub;
        $this->assertMatchesRegularExpression('/<!--/', $stub->getVimeoData(null));
        $this->assertMatchesRegularExpression('/<!--/', $stub->getVimeoData(''));
        // TODO: Test for a malformed data set with a missing Vimeo->Pictures->sizes array
    }

    /**
     * It should catch Vimeo API Errors
     *
     * @runInSeparateProcess
     * @preserveGlobalState disabled
     */
    public function testVimeoAPIError()
    {
        define('WP_DEBUG', true);

        $stub = $this->getMockBuilder('ideasonpurpose\VimeoEmbed')
            ->disableOriginalConstructor()
            ->setMethodsExcept(['apiGet', 'throwError'])
            ->getMock();

        $mock = new MockHandler([
            new Response(200, [], '{"id": 123, "name":"test video"}'),
            new Response(500, [], 'testing'),
            new RequestException('Error Communicating with Server', new Request('GET', 'test')),
        ]);
        $handler = HandlerStack::create($mock);
        $stub->client = new Client(['handler' => $handler]);
        $stub->token = '123456789abcdefg';
        $this->assertIsObject($stub->apiGet('returns an object'));

        $this->expectException(\Exception::class);
        $stub->apiGet('500 error');

        $this->expectException(\Exception::class);
        $stub->apiGet('throw exception');
    }

    /**
     * it should create a video tag embed with the loop attribute
     * it should create a video tag embed with the autoplay attribute
     * it should include muted and playsinline with the autoplay attribute
     */
    public function testHTML5Attributes()
    {
        $stub = $this->getMockBuilder('ideasonpurpose\VimeoEmbed')
            ->disableOriginalConstructor()
            // ->setMethods(['getVimeoData', 'divStart'])
            ->onlyMethods(['getVimeoData', 'divStart'])
            ->getMock();

        $stub->method('getVimeoData')->willReturn(
            (object) [
                'files' => 'files array',
                'pictures' => (object) ['sizes' => 'pictures array'],
            ]
        );

        $this->assertMatchesRegularExpression(
            '/autoplay/',
            $stub->embed(1234, ['autoplay' => true])
        );
        $this->assertDoesNotMatchRegularExpression(
            '/autoplay/',
            $stub->embed(1234, ['autoplay' => false])
        );
        $this->assertMatchesRegularExpression('/muted/', $stub->embed(1234, ['autoplay' => true]));
        $this->assertMatchesRegularExpression(
            '/playsinline/',
            $stub->embed(1234, ['autoplay' => true])
        );

        $this->assertMatchesRegularExpression('/loop/', $stub->embed(1234, ['loop' => true]));
        $this->assertDoesNotMatchRegularExpression('/loop/', $stub->embed(1234, ['loop' => false]));
    }

    public function testDivStart()
    {
        global $stub;
        $fakeData = (object) [
            'width' => 16,
            'height' => 9,
            'embed' => (object) ['html' => 'html body'],
        ];
        $div = $stub->divStart($fakeData);
        $this->assertMatchesRegularExpression('/<style>/', $div);
        $this->assertMatchesRegularExpression('/<div id="vimeo-embed/', $div);
    }

    public function testDivStartNoStyle()
    {
        global $stub;
        $fakeData = (object) [
            'width' => 16,
            'height' => 9,
            'embed' => (object) ['html' => 'html body'],
        ];
        $div = $stub->divStart($fakeData, false);
        $this->assertDoesNotMatchRegularExpression('/<style>/', $div);
        $this->assertDoesNotMatchRegularExpression('/<div id="vimeo-embed/', $div);
        $this->assertMatchesRegularExpression('/<div class="embed-container/', $div);
    }
}


src/js/ekko-lightbox.js
/**
 * This is a temporary tweak to the Ekko-Lightbox JS module to fix hard-coded
 * Vimeo modal widths. Line 387 was changed.
 *
 * Waiting for https://github.com/ashleydw/lightbox/issues/244 to be merged.
 * Will be replaced with the CDN copy once #244 is merged.
 *
 * Ref:
 * https://github.com/ashleydw/lightbox/issues/247
 * https://github.com/ashleydw/lightbox/pull/239
 */

const Lightbox = (($) => {

  const NAME = 'ekkoLightbox'
  const JQUERY_NO_CONFLICT = $.fn[NAME]

  const Default = {
    title: '',
    footer: '',
    showArrows: true, //display the left / right arrows or not
    type: null, //force the lightbox into image / youtube mode. if null, or not image|youtube|vimeo; detect it
    alwaysShowClose: false, //always show the close button, even if there is no title
    loadingMessage: '<div class="ekko-lightbox-loader"><div><div></div><div></div></div></div>', // http://tobiasahlin.com/spinkit/
    leftArrow: '<span>&#10094;</span>',
    rightArrow: '<span>&#10095;</span>',
    strings: {
      close: 'Close',
      fail: 'Failed to load image:',
      type: 'Could not detect remote target type. Force the type using data-type',
    },
    doc: document, // if in an iframe can specify top.document
    onShow() {},
    onShown() {},
    onHide() {},
    onHidden() {},
    onNavigate() {},
    onContentLoaded() {}
  }

  class Lightbox {

    /**

      Class properties:

     _$element: null -> the <a> element currently being displayed
     _$modal: The bootstrap modal generated
        _$modalDialog: The .modal-dialog
        _$modalContent: The .modal-content
        _$modalBody: The .modal-body
        _$modalHeader: The .modal-header
        _$modalFooter: The .modal-footer
     _$lightboxContainerOne: Container of the first lightbox element
     _$lightboxContainerTwo: Container of the second lightbox element
     _$lightboxBody: First element in the container
     _$modalArrows: The overlayed arrows container

     _$galleryItems: Other <a>'s available for this gallery
     _galleryName: Name of the current data('gallery') showing
     _galleryIndex: The current index of the _$galleryItems being shown

     _config: {} the options for the modal
     _modalId: unique id for the current lightbox
     _padding / _border: CSS properties for the modal container; these are used to calculate the available space for the content

     */

    static get Default() {
      return Default
    }

    constructor($element, config) {
      this._config = $.extend({}, Default, config)
      this._$modalArrows = null
      this._galleryIndex = 0
      this._galleryName = null
      this._padding = null
      this._border = null
      this._titleIsShown = false
      this._footerIsShown = false
      this._wantedWidth = 0
      this._wantedHeight = 0
      this._modalId = `ekkoLightbox-${Math.floor((Math.random() * 1000) + 1)}`;
      this._$element = $element instanceof jQuery ? $element : $($element)

      this._isBootstrap3 = $.fn.modal.Constructor.VERSION[0] == 3;

      let h4 = `<h4 class="modal-title">${this._config.title || "&nbsp;"}</h4>`;
      let btn = `<button type="button" class="close" data-dismiss="modal" aria-label="${this._config.strings.close}"><span aria-hidden="true">&times;</span></button>`;

      let header = `<div class="modal-header"${this._config.title || this._config.alwaysShowClose ? '' : ' style="display:none"'}>`+(this._isBootstrap3 ? btn+h4 : h4+btn)+`</div>`;
      let footer = `<div class="modal-footer"${this._config.footer ? '' : ' style="display:none"'}>${this._config.footer || "&nbsp;"}</div>`;
      let body = '<div class="modal-body"><div class="ekko-lightbox-container"><div class="ekko-lightbox-item fade in show"></div><div class="ekko-lightbox-item fade"></div></div></div>'
      let dialog = `<div class="modal-dialog" role="document"><div class="modal-content">${header}${body}${footer}</div></div>`
      $(this._config.doc.body).append(`<div id="${this._modalId}" class="ekko-lightbox modal fade" tabindex="-1" tabindex="-1" role="dialog" aria-hidden="true">${dialog}</div>`)

      this._$modal = $(`#${this._modalId}`, this._config.doc)
      this._$modalDialog = this._$modal.find('.modal-dialog').first()
      this._$modalContent = this._$modal.find('.modal-content').first()
      this._$modalBody = this._$modal.find('.modal-body').first()
      this._$modalHeader = this._$modal.find('.modal-header').first()
      this._$modalFooter = this._$modal.find('.modal-footer').first()

      this._$lightboxContainer = this._$modalBody.find('.ekko-lightbox-container').first()
      this._$lightboxBodyOne = this._$lightboxContainer.find('> div:first-child').first()
      this._$lightboxBodyTwo = this._$lightboxContainer.find('> div:last-child').first()

      this._border = this._calculateBorders()
      this._padding = this._calculatePadding()

      this._galleryName = this._$element.data('gallery')
      if (this._galleryName) {
        this._$galleryItems = $(document.body).find(`*[data-gallery="${this._galleryName}"]`)
        this._galleryIndex = this._$galleryItems.index(this._$element)
        $(document).on('keydown.ekkoLightbox', this._navigationalBinder.bind(this))

        // add the directional arrows to the modal
        if (this._config.showArrows && this._$galleryItems.length > 1) {
          this._$lightboxContainer.append(`<div class="ekko-lightbox-nav-overlay"><a href="#">${this._config.leftArrow}</a><a href="#">${this._config.rightArrow}</a></div>`)
          this._$modalArrows = this._$lightboxContainer.find('div.ekko-lightbox-nav-overlay').first()
          this._$lightboxContainer.on('click', 'a:first-child', event => {
            event.preventDefault()
            return this.navigateLeft()
          })
          this._$lightboxContainer.on('click', 'a:last-child', event => {
            event.preventDefault()
            return this.navigateRight()
          })
        }
      }

      this._$modal
      .on('show.bs.modal', this._config.onShow.bind(this))
      .on('shown.bs.modal', () => {
        this._toggleLoading(true)
        this._handle()
        return this._config.onShown.call(this)
      })
      .on('hide.bs.modal', this._config.onHide.bind(this))
      .on('hidden.bs.modal', () => {
        if (this._galleryName) {
          $(document).off('keydown.ekkoLightbox')
          $(window).off('resize.ekkoLightbox')
        }
        this._$modal.remove()
        return this._config.onHidden.call(this)
      })
      .modal(this._config)

      $(window).on('resize.ekkoLightbox', () => {
        this._resize(this._wantedWidth, this._wantedHeight)
      })
    }

    element() {
      return this._$element;
    }

    modal() {
      return this._$modal;
    }

    navigateTo(index) {

      if (index < 0 || index > this._$galleryItems.length-1)
        return this

      this._galleryIndex = index

      this._$element = $(this._$galleryItems.get(this._galleryIndex))
      this._handle();
    }

    navigateLeft() {

      if (this._$galleryItems.length === 1)
        return

      if (this._galleryIndex === 0)
        this._galleryIndex = this._$galleryItems.length - 1
      else //circular
        this._galleryIndex--

      this._config.onNavigate.call(this, 'left', this._galleryIndex)
      return this.navigateTo(this._galleryIndex)
    }

    navigateRight() {

      if (this._$galleryItems.length === 1)
        return

      if (this._galleryIndex === this._$galleryItems.length - 1)
        this._galleryIndex = 0
      else //circular
        this._galleryIndex++

      this._config.onNavigate.call(this, 'right', this._galleryIndex)
      return this.navigateTo(this._galleryIndex)
    }

    close() {
      return this._$modal.modal('hide');
    }

    // helper private methods
    _navigationalBinder(event) {
      event = event || window.event;
      if (event.keyCode === 39)
        return this.navigateRight()
      if (event.keyCode === 37)
        return this.navigateLeft()
    }

    // type detection private methods
    _detectRemoteType(src, type) {

      type = type || false;

      if(!type && this._isImage(src))
        type = 'image';
      if(!type && this._getYoutubeId(src))
        type = 'youtube';
      if(!type && this._getVimeoId(src))
        type = 'vimeo';
      if(!type && this._getInstagramId(src))
        type = 'instagram';

      if(!type || ['image', 'youtube', 'vimeo', 'instagram', 'video', 'url'].indexOf(type) < 0)
        type = 'url';

      return type;
    }

    _isImage(string) {
      return string && string.match(/(^data:image\/.*,)|(\.(jp(e|g|eg)|gif|png|bmp|webp|svg)((\?|#).*)?$)/i)
    }

    _containerToUse() {
      // if currently showing an image, fade it out and remove
      let $toUse = this._$lightboxBodyTwo
      let $current = this._$lightboxBodyOne

      if(this._$lightboxBodyTwo.hasClass('in')) {
        $toUse = this._$lightboxBodyOne
        $current = this._$lightboxBodyTwo
      }

      $current.removeClass('in show')
      setTimeout(() => {
        if(!this._$lightboxBodyTwo.hasClass('in'))
          this._$lightboxBodyTwo.empty()
        if(!this._$lightboxBodyOne.hasClass('in'))
          this._$lightboxBodyOne.empty()
      }, 500)

      $toUse.addClass('in show')
      return $toUse
    }

    _handle() {

      let $toUse = this._containerToUse()
      this._updateTitleAndFooter()

      let currentRemote = this._$element.attr('data-remote') || this._$element.attr('href')
      let currentType = this._detectRemoteType(currentRemote, this._$element.attr('data-type') || false)

      if(['image', 'youtube', 'vimeo', 'instagram', 'video', 'url'].indexOf(currentType) < 0)
        return this._error(this._config.strings.type)

      switch(currentType) {
        case 'image':
          this._preloadImage(currentRemote, $toUse)
          this._preloadImageByIndex(this._galleryIndex, 3)
          break;
        case 'youtube':
          this._showYoutubeVideo(currentRemote, $toUse);
          break;
        case 'vimeo':
          this._showVimeoVideo(this._getVimeoId(currentRemote), $toUse);
          break;
        case 'instagram':
          this._showInstagramVideo(this._getInstagramId(currentRemote), $toUse);
          break;
        case 'video':
          this._showHtml5Video(currentRemote, $toUse);
          break;
        default: // url
          this._loadRemoteContent(currentRemote, $toUse);
          break;
      }

      return this;
    }

    _getYoutubeId(string) {
      if(!string)
        return false;
      let matches = string.match(/^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=)([^#\&\?]*).*/)
      return (matches && matches[2].length === 11) ? matches[2] : false
    }

    _getVimeoId(string) {
      return string && string.indexOf('vimeo') > 0 ? string : false
    }

    _getInstagramId(string) {
      return string && string.indexOf('instagram') > 0 ? string : false
    }

    // layout private methods
    _toggleLoading(show) {
      show = show || false
      if(show) {
        this._$modalDialog.css('display', 'none')
        this._$modal.removeClass('in show')
        $('.modal-backdrop').append(this._config.loadingMessage)
      }
      else {
        this._$modalDialog.css('display', 'block')
        this._$modal.addClass('in show')
        $('.modal-backdrop').find('.ekko-lightbox-loader').remove()
      }
      return this;
    }

    _calculateBorders() {
      return {
        top: this._totalCssByAttribute('border-top-width'),
        right: this._totalCssByAttribute('border-right-width'),
        bottom: this._totalCssByAttribute('border-bottom-width'),
        left: this._totalCssByAttribute('border-left-width'),
      }
    }

    _calculatePadding() {
      return {
        top: this._totalCssByAttribute('padding-top'),
        right: this._totalCssByAttribute('padding-right'),
        bottom: this._totalCssByAttribute('padding-bottom'),
        left: this._totalCssByAttribute('padding-left'),
      }
    }

    _totalCssByAttribute(attribute) {
      return parseInt(this._$modalDialog.css(attribute), 10) +
        parseInt(this._$modalContent.css(attribute), 10) +
        parseInt(this._$modalBody.css(attribute), 10)
    }

    _updateTitleAndFooter() {
      let title = this._$element.data('title') || ""
      let caption = this._$element.data('footer') || ""

      this._titleIsShown = false
      if (title || this._config.alwaysShowClose) {
        this._titleIsShown = true
        this._$modalHeader.css('display', '').find('.modal-title').html(title || "&nbsp;")
      }
      else
        this._$modalHeader.css('display', 'none')

      this._footerIsShown = false
      if (caption) {
        this._footerIsShown = true
        this._$modalFooter.css('display', '').html(caption)
      }
      else
        this._$modalFooter.css('display', 'none')

      return this;
    }

    _showYoutubeVideo(remote, $containerForElement) {
      let id = this._getYoutubeId(remote)
      let query = remote.indexOf('&') > 0 ? remote.substr(remote.indexOf('&')) : ''
      let width = this._$element.data('width') || 560
      let height = this._$element.data('height') ||  width / ( 560/315 )
      return this._showVideoIframe(
        `//www.youtube.com/embed/${id}?badge=0&autoplay=1&html5=1${query}`,
        width,
        height,
        $containerForElement
      );
    }

    _showVimeoVideo(id, $containerForElement) {
      let width = this._$element.data('width') || 500
      let height = this._$element.data('height') ||  width / ( 560/315 )
      return this._showVideoIframe(id + '?autoplay=1', width, height, $containerForElement)
    }

    _showInstagramVideo(id, $containerForElement) {
      // instagram load their content into iframe's so this can be put straight into the element
      let width = this._$element.data('width') || 612
      let height = width + 80;
      id = id.substr(-1) !== '/' ? id + '/' : id; // ensure id has trailing slash
      $containerForElement.html(`<iframe width="${width}" height="${height}" src="${id}embed/" frameborder="0" allowfullscreen></iframe>`);
      this._resize(width, height);
      this._config.onContentLoaded.call(this);
      if (this._$modalArrows) //hide the arrows when showing video
        this._$modalArrows.css('display', 'none');
      this._toggleLoading(false);
      return this;
    }

    _showVideoIframe(url, width, height, $containerForElement) { // should be used for videos only. for remote content use loadRemoteContent (data-type=url)
      height = height || width; // default to square
      $containerForElement.html(`<div class="embed-responsive embed-responsive-16by9"><iframe width="${width}" height="${height}" src="${url}" frameborder="0" allowfullscreen class="embed-responsive-item"></iframe></div>`);
      this._resize(width, height);
      this._config.onContentLoaded.call(this);
      if (this._$modalArrows)
        this._$modalArrows.css('display', 'none'); //hide the arrows when showing video
      this._toggleLoading(false);
      return this;
    }

    _showHtml5Video(url, $containerForElement) { // should be used for videos only. for remote content use loadRemoteContent (data-type=url)
      let width = this._$element.data('width') || 560
      let height = this._$element.data('height') ||  width / ( 560/315 )
      $containerForElement.html(`<div class="embed-responsive embed-responsive-16by9"><video width="${width}" height="${height}" src="${url}" preload="auto" autoplay controls class="embed-responsive-item"></video></div>`);
      this._resize(width, height);
      this._config.onContentLoaded.call(this);
      if (this._$modalArrows)
        this._$modalArrows.css('display', 'none'); //hide the arrows when showing video
      this._toggleLoading(false);
      return this;
    }

    _loadRemoteContent(url, $containerForElement) {
      let width = this._$element.data('width') || 560;
      let height = this._$element.data('height') || 560;

      let disableExternalCheck = this._$element.data('disableExternalCheck') || false;
      this._toggleLoading(false);

      // external urls are loading into an iframe
      // local ajax can be loaded into the container itself
      if (!disableExternalCheck && !this._isExternal(url)) {
        $containerForElement.load(url, $.proxy(() => {
          return this._$element.trigger('loaded.bs.modal');l
        }));

      } else {
        $containerForElement.html(`<iframe src="${url}" frameborder="0" allowfullscreen></iframe>`);
        this._config.onContentLoaded.call(this);
      }

      if (this._$modalArrows) //hide the arrows when remote content
        this._$modalArrows.css('display', 'none')

      this._resize(width, height);
      return this;
    }

    _isExternal(url) {
      let match = url.match(/^([^:\/?#]+:)?(?:\/\/([^\/?#]*))?([^?#]+)?(\?[^#]*)?(#.*)?/);
      if (typeof match[1] === "string" && match[1].length > 0 && match[1].toLowerCase() !== location.protocol)
        return true;

      if (typeof match[2] === "string" && match[2].length > 0 && match[2].replace(new RegExp(`:(${{
          "http:": 80,
          "https:": 443
        }[location.protocol]})?$`), "") !== location.host)
        return true;

      return false;
    }

    _error( message ) {
      console.error(message);
      this._containerToUse().html(message);
      this._resize(300, 300);
      return this;
    }

    _preloadImageByIndex(startIndex, numberOfTimes) {

      if(!this._$galleryItems)
        return;

      let next = $(this._$galleryItems.get(startIndex), false)
      if(typeof next == 'undefined')
        return

      let src = next.attr('data-remote') || next.attr('href')
      if (next.attr('data-type') === 'image' || this._isImage(src))
        this._preloadImage(src, false)

      if(numberOfTimes > 0)
        return this._preloadImageByIndex(startIndex + 1, numberOfTimes-1);
    }

    _preloadImage( src, $containerForImage) {

      $containerForImage = $containerForImage || false

      let img = new Image();
      if ($containerForImage) {

        // if loading takes > 200ms show a loader
        let loadingTimeout = setTimeout(() => {
          $containerForImage.append(this._config.loadingMessage)
        }, 200)

        img.onload = () => {
          if(loadingTimeout)
            clearTimeout(loadingTimeout)
          loadingTimeout = null;
          let image = $('<img />');
          image.attr('src', img.src);
          image.addClass('img-fluid');

          // backward compatibility for bootstrap v3
          image.css('width', '100%');

          $containerForImage.html(image);
          if (this._$modalArrows)
            this._$modalArrows.css('display', '') // remove display to default to css property

          this._resize(img.width, img.height);
          this._toggleLoading(false);
          return this._config.onContentLoaded.call(this);
        };
        img.onerror = () => {
          this._toggleLoading(false);
          return this._error(this._config.strings.fail+`  ${src}`);
        };
      }

      img.src = src;
      return img;
    }

    _resize( width, height ) {

      height = height || width
      this._wantedWidth = width
      this._wantedHeight = height

      // if width > the available space, scale down the expected width and height
      let widthBorderAndPadding = this._padding.left + this._padding.right + this._border.left + this._border.right
      let maxWidth = Math.min(width + widthBorderAndPadding, this._config.doc.body.clientWidth)
      if((width + widthBorderAndPadding) > maxWidth) {
        height = ((maxWidth - widthBorderAndPadding) /  width) * height
        width = maxWidth
      } else
        width = (width + widthBorderAndPadding)

      let headerHeight = 0,
        footerHeight = 0

      // as the resize is performed the modal is show, the calculate might fail
      // if so, default to the default sizes
      if (this._footerIsShown)
        footerHeight = this._$modalFooter.outerHeight(true) || 55

      if (this._titleIsShown)
        headerHeight = this._$modalHeader.outerHeight(true) || 67

      let borderPadding = this._padding.top + this._padding.bottom + this._border.bottom + this._border.top

      //calculated each time as resizing the window can cause them to change due to Bootstraps fluid margins
      let margins = parseFloat(this._$modalDialog.css('margin-top')) + parseFloat(this._$modalDialog.css('margin-bottom'));

      let maxHeight = Math.min(height, $(window).height() - borderPadding - margins - headerHeight - footerHeight);
      if(height > maxHeight) {
        // if height > the available height, scale down the width
        let factor = Math.min(maxHeight / height, 1);
        width = Math.ceil(factor * width);
      }

      this._$lightboxContainer.css('height', maxHeight)
      this._$modalDialog.css('width', 'auto') .css('maxWidth', width);

      if (!this._isBootstrap3) {
        // v4 method is mistakenly protected
        let modal = this._$modal.data('bs.modal');
        if (modal)
          modal._handleUpdate();
      } else {
        let modal = this._$modal.data('bs.modal');
        if (modal)
          modal.handleUpdate();
      }
      return this;
    }

    static _jQueryInterface(config) {
      config = config || {}
      return this.each(() => {
        let $this = $(this)
        let _config = $.extend(
          {},
          Lightbox.Default,
          $this.data(),
          typeof config === 'object' && config
        )

        new Lightbox(this, _config)
      })
    }
  }



  $.fn[NAME]             = Lightbox._jQueryInterface
  $.fn[NAME].Constructor = Lightbox
  $.fn[NAME].noConflict  = () => {
    $.fn[NAME] = JQUERY_NO_CONFLICT
    return Lightbox._jQueryInterface
  }

  return Lightbox

})(jQuery)

export default Lightbox
